### PR TITLE
New ModelTSO() class (ATLAS and PHOENIX stellar models support)

### PR DESCRIPTION
**Pull request by nespinoza**
Hi Joe,

 In this PR, I implemented the `ModelTSO()` class in `awesim.py` which generalizes the `TestTSO` and `BlackbodyTSO` classes by allowing the user to use an ATLAS or PHOENIX stellar model as input fluxes. The models are automatically downloaded based on the user inputs (the code right now selects the closest ATLAS or PHOENIX stellar model to the input stellar parameters --- perhaps in the future we could provide interpolations, but this is not too critical I think).

 On top of this, the class also allows you to scale down the flux of the stellar model in order to emulate the flux observed at Earth using an input J-magnitude. This is done using Vega as the absolute flux calibrator --- I took the spectrum of Vega from the `CALSPEC` library (http://www.stsci.edu/hst/instrumentation/reference-data-for-calibration-and-tools/astronomical-catalogs/calspec). I compared these scaled fluxes to the spectrum under `awesimsoss/files/scaled_spectrum.txt` and the match is excellent. With this PR, apart from the modifications to the `awesim.py` and the tests scripts, I'm also pushing the Vega spectrum and the J-band transmission filter, which is needed for these calculations.

 I followed all the guidelines for contributions *except* for the fact that I didn't create a different branch for this (my bad --- sorry!). I will be careful to do this in the future --- I presume this is no big deal in this particular case but, if it is, let me know.

 Questions and/or comments, I'm only one Github comment or email away!

Néstor

### CODE REVIEW

**Comment by hover2pi** [comment]
@nespinoza Also, you had a bunch of `try/except` statements in the `ModelTSO` class that will need to be tested in the `test_awesim.py` module. You have the basic initialization tests in there mimicking the tests for the other TSO child classes, but the `ModelTSO` child class is significantly more complicated. The PR checks will fail if the test coverage decreases, unfortunately. We could always add these tests later too. I just wanted to let you know. Thanks!

**Comment by nespinoza** [comment]
Hi Joe! 

 Thanks for the feedback --- I'll take a look a the formating issues you mentioned and get back to this in the next few weeks!

N.

File: awesimsoss/awesim.py [hunk 1]
```diff
@@ -37,6 +37,16 @@
 from . import generate_darks as gd
 from . import make_trace as mt
 
+# Some extra imports for PHOENIX model downloads:
+import zipfile
```
[the remark this answers is not in the document]
Comment (nespinoza) [reply awesimsoss/awesim.py]: All right, here we go. First this one, addressed in: https://github.com/spacetelescope/awesimsoss/pull/61/commits/763a66e5b8d18312f6c26dae3bf43523e2293ac1

File: awesimsoss/awesim.py [hunk 2]
```diff
@@ -1367,3 +1377,369 @@ def __init__(self, ngrps=2, nints=2, teff=1800, filter='CLEAR', subarray='SUBSTR
...
+        ZP = -0.001 + 2.5*np.log10(vega_weighted_flux)
+        # Now compute (inverse?) bolometric correction for target star. For this, compute same integral as for vega, but for target:
+        target_weighted_flux = self.spec_integral(np.array(w)*1e4, f, np.array(wT)*1e4, TT)
+        # Get scaling factor for target spectrum (this ommits any extinction):
+        scaling_factor = 10**(-((jmag + 2.5*np.log10(target_weighted_flux) - ZP)/2.5))
+        # Return scaled spectrum:
+        return f*scaling_factor
+
+    """Generate an object with an ATLAS or PHOENIX stellar spectrum"""
+    def __init__(self, ngrps=2, nints=2, teff=5700.0, logg = 4.0, feh = 0.0, alpha = 0.0, jmag = 9.0, stellar_model = 'ATLAS', filter='CLEAR', subarray='SUBSTRIP256', run=True, add_planet=False, scale=1., **kwargs):
```
[the remark this answers is not in the document]
Comment (nespinoza) [reply awesimsoss/awesim.py]: Done, addressed in https://github.com/spacetelescope/awesimsoss/pull/61/commits/8707cf284f06e11af5a1c0ecd59d2c40a8c670e8

File: awesimsoss/awesim.py [hunk 3]
```diff
@@ -1367,3 +1377,369 @@ def __init__(self, ngrps=2, nints=2, teff=1800, filter='CLEAR', subarray='SUBSTR
...
+            print('\t Warning: the input stellar log-gravity is outside the {0:}-{1:} model range of ATLAS models for {2:} and Teff {3:}.'.format(\
+                   np.min(real_possible_loggs),np.max(real_possible_loggs),atlas_met,atlas_teff))
+
+        # Change units in order to match what is expected by the TSO modules:
+        wav = (w * q.angstrom).to(q.um)
+        flux = f * q.erg/q.s/q.cm**2/q.AA
+        return wav, flux
+
+    def get_resolution(self, w, f):
+        eff_wav = np.sum(w*f)/np.sum(f)
```
[the remark this answers is not in the document]
Comment (nespinoza) [reply awesimsoss/awesim.py]: Done in https://github.com/spacetelescope/awesimsoss/pull/61/commits/90f18227a23679f1ee80e378ba7f67157c9b883b

File: awesimsoss/awesim.py [hunk 4]
```diff
@@ -1367,3 +1377,369 @@ def __init__(self, ngrps=2, nints=2, teff=1800, filter='CLEAR', subarray='SUBSTR
...
+    def get_vega(self):
+        """
+        This functions reads in the spectrum of Vega (Alpha Lyr) from CALSPEC
+        """
+        data = fits.getdata(resource_filename('awesimsoss', 'files/alpha_lyr_stis_009.fits'),header=False)
+        # Wavelength is in Angstroms, convert to microns to match the get_phoenix_model function. 
+        # Flux is in Flambda (same as Phoenix; i.e., erg/s/cm2/A):
+        return (data['WAVELENGTH']*q.angstrom).to(q.um),data['FLUX']*(q.erg/q.s/q.cm**2/q.AA) 
+
+    def read_phoenix_list(self, phoenix_model_list):
```
[the remark this answers is not in the document]
Comment (nespinoza) [reply awesimsoss/awesim.py]: Done in https://github.com/spacetelescope/awesimsoss/pull/61/commits/90f18227a23679f1ee80e378ba7f67157c9b883b

File: awesimsoss/awesim.py [hunk 5]
```diff
@@ -1367,3 +1377,369 @@ def __init__(self, ngrps=2, nints=2, teff=1800, filter='CLEAR', subarray='SUBSTR
...
+        else:
+            model_alpha = np.abs(model_alpha)
+        # Create the folder name
+        if alpha == 0.0:
+            fname = 'ftp://phoenix.astro.physik.uni-goettingen.de/HiResFITS/PHOENIX-ACES-AGSS-COND-2011/Z{0:}{1:.1f}/'.format(met_sign,model_metallicity)
+        else:
+            fname = 'ftp://phoenix.astro.physik.uni-goettingen.de/HiResFITS/PHOENIX-ACES-AGSS-COND-2011/Z{0:}{1:.1f}.Alpha={2:}{3:.2f}/'.format(met_sign,model_metallicity,alpha_sign,model_alpha)
+        return fname
+
+    def download(self, url, fname):
```
[the remark this answers is not in the document]
Comment (nespinoza) [reply awesimsoss/awesim.py]: Done in https://github.com/spacetelescope/awesimsoss/pull/61/commits/90f18227a23679f1ee80e378ba7f67157c9b883b

File: awesimsoss/awesim.py [hunk 6]
```diff
@@ -1367,3 +1377,369 @@ def __init__(self, ngrps=2, nints=2, teff=1800, filter='CLEAR', subarray='SUBSTR
...
+        # Define the sign before the filename, obtain absolute value if needed:
+        if model_metallicity >= 0.0:
+            met_sign = 'p'
+        else:
+            model_metallicity = np.abs(model_metallicity)
+        model_metallicity = ''.join(str(model_metallicity).split('.'))
+        fname = 'https://archive.stsci.edu/hlsps/reference-atlases/cdbs/grid/ck04models/ck{0:}{1:}/'.format(met_sign,model_metallicity)
+        return fname
+
+    def get_phoenix_folder(self, feh, alpha):
```
[the remark this answers is not in the document]
Comment (nespinoza) [reply awesimsoss/awesim.py]: Done in https://github.com/spacetelescope/awesimsoss/pull/61/commits/90f18227a23679f1ee80e378ba7f67157c9b883b

File: awesimsoss/awesim.py [hunk 7]
```diff
@@ -1367,3 +1377,369 @@ def __init__(self, ngrps=2, nints=2, teff=1800, filter='CLEAR', subarray='SUBSTR
...
+    def closest_value(self, input_value, possible_values):
+        """
+        This function calculates, given an input_value and an array of possible_values, 
+        the closest value to input_value in the array.
+        """
+        distance = np.abs(possible_values - input_value)
+        idx = np.where(distance == np.min(distance))[0]
+        return possible_values[idx[0]]
+
+    def get_atlas_folder(self, feh):
```
[the remark this answers is not in the document]
Comment (nespinoza) [reply awesimsoss/awesim.py]: Done in https://github.com/spacetelescope/awesimsoss/pull/61/commits/90f18227a23679f1ee80e378ba7f67157c9b883b

File: awesimsoss/awesim.py [hunk 8]
```diff
@@ -1367,3 +1377,369 @@ def __init__(self, ngrps=2, nints=2, teff=1800, filter='CLEAR', subarray='SUBSTR
         # Run the simulation
         if run:
             self.simulate()
+
+class ModelTSO(TSO):
+    def closest_value(self, input_value, possible_values):
```
[the remark this answers is not in the document]
Comment (nespinoza) [reply awesimsoss/awesim.py]: Done in https://github.com/spacetelescope/awesimsoss/pull/61/commits/90f18227a23679f1ee80e378ba7f67157c9b883b

File: awesimsoss/awesim.py [hunk 9]
```diff
@@ -1367,3 +1377,369 @@ def __init__(self, ngrps=2, nints=2, teff=1800, filter='CLEAR', subarray='SUBSTR
...
+                fname = line.split()[-1]
+                teff, logg = fname.split('-')[:2]
+                fnames = np.append(fnames, fname)
+                teffs = np.append(teffs, np.double(teff[3:]))
+                loggs = np.append(loggs, np.double(logg))
+            else:
+                break
+        return fnames, teffs, loggs
+    
+    def get_phoenix_model(self, feh, alpha, teff, logg):
```
[the remark this answers is not in the document]
Comment (nespinoza) [reply awesimsoss/awesim.py]: Done in https://github.com/spacetelescope/awesimsoss/pull/61/commits/90f18227a23679f1ee80e378ba7f67157c9b883b

File: awesimsoss/awesim.py [hunk 10]
```diff
@@ -1367,3 +1377,369 @@ def __init__(self, ngrps=2, nints=2, teff=1800, filter='CLEAR', subarray='SUBSTR
...
+        # Once we have the file, simply extract the data:
+        print('\t Using the {0:} PHOENIX model (Teff {1:}, logg {2:}).'.format(phoenix_model, phoenix_teff, phoenix_logg))
+        flux = fits.getdata(model_folder_path+phoenix_model,header=False)
+
+        # Change units in order to match what is expected by the TSO modules:
+        wav = (wavelengths * q.angstrom).to(q.um)
+        flux = (flux * (q.erg/q.s/q.cm**2/q.cm)).to(q.erg/q.s/q.cm**2/q.AA)
+        return wav, flux
+
+    def get_atlas_model(self, feh, teff, logg):
```
[the remark this answers is not in the document]
Comment (nespinoza) [reply awesimsoss/awesim.py]: Done in https://github.com/spacetelescope/awesimsoss/pull/61/commits/90f18227a23679f1ee80e378ba7f67157c9b883b

File: awesimsoss/awesim.py [hunk 11]
```diff
@@ -1367,3 +1377,369 @@ def __init__(self, ngrps=2, nints=2, teff=1800, filter='CLEAR', subarray='SUBSTR
...
+        wav = (w * q.angstrom).to(q.um)
+        flux = f * q.erg/q.s/q.cm**2/q.AA
+        return wav, flux
+
+    def get_resolution(self, w, f):
+        eff_wav = np.sum(w*f)/np.sum(f)
+        delta_wav = np.median(np.abs(np.diff(w)))
+        return eff_wav/delta_wav
+
+    def spec_integral(self, input_w, input_f, wT, TT):
```
[the remark this answers is not in the document]
Comment (nespinoza) [reply awesimsoss/awesim.py]: Done in https://github.com/spacetelescope/awesimsoss/pull/61/commits/90f18227a23679f1ee80e378ba7f67157c9b883b

File: awesimsoss/awesim.py [hunk 12]
```diff
@@ -1367,3 +1377,369 @@ def __init__(self, ngrps=2, nints=2, teff=1800, filter='CLEAR', subarray='SUBSTR
...
+            f = ndimage.gaussian_filter(input_f[idx],int(np.double(len(idx))/np.double(len(wT))))
+            w = input_w[idx]
+        else:
+            w, f = input_w, input_f
+        interp_spectra = interpolate.interp1d(w,f)
+        numerator = np.trapz(wT*interp_spectra(wT)*TT, x = wT)
+        denominator = np.trapz(wT*TT, x = wT)
+        return numerator/denominator
+
+    def scale_spectrum(self, w, f, jmag):
```
[the remark this answers is not in the document]
Comment (nespinoza) [reply awesimsoss/awesim.py]: Done in https://github.com/spacetelescope/awesimsoss/pull/61/commits/90f18227a23679f1ee80e378ba7f67157c9b883b

**Comment by nespinoza** [comment]
Hey @hover2pi!

I finally came back to this. I gave my best to implement all the changes you mentioned (which, by the way, was quite pedagogical for me!), including pushing the `__init__.py` (here: https://github.com/spacetelescope/awesimsoss/pull/61/commits/e7ec64a7f55440a93619a133e3e261d36d121d0a), which I had changed but not committed (duh!).

I'm not sure to which try/except statements you are referring in the `ModelTSO` class in your latest comment, but as you say, we can figure out the full-blown set of tests later. For now, I think this should be "good enough" I believe, but if you believe otherwise do let me know --- I absolutely trust much more your judgement on coding practices than mine.

Thanks for reviewing so thoroughly this PR!

Néstor

**Comment by hover2pi** [comment]
Sorry it took so long to get back to this @nespinoza ! I merged the changes. Thanks!

**Comment by nespinoza** [comment]
Hey no worries, glad it got merged!

**Pull request merged**